Allow reloading of search time analyzers

cbuescher opened the pull request:
Currently changing resources (like dictionaries, synonym files etc...) of search
time analyzers is only possible by closing an index, changing the underlying
resource (e.g. synonym files) and then re-opening the index for the change to
take effect.

This PR adds a new API endpoint that allows triggering reloading of certain
analysis resources (currently token filters) that will then pick up changes in
underlying file resources. To achieve this we introduce a new type of custom
analyzer (ReloadableCustomAnalyzer) that uses a ReuseStrategy that allows
swapping out analysis components. Custom analyzers that contain filters that are
markes as "updateable" will automatically choose this implementation. This PR
also adds this capability to synonym token filters for use in search time
analyzers.

Relates to #29051

Commit:
Allow reloading of search time analyzers (#42888)

Currently changing resources (like dictionaries, synonym files etc...) of search
time analyzers is only possible by closing an index, changing the underlying
resource (e.g. synonym files) and then re-opening the index for the change to
take effect.

This PR adds a new API endpoint that allows triggering reloading of certain
analysis resources (currently token filters) that will then pick up changes in
underlying file resources. To achieve this we introduce a new type of custom
analyzer (ReloadableCustomAnalyzer) that uses a ReuseStrategy that allows
swapping out analysis components. Custom analyzers that contain filters that are
markes as "updateable" will automatically choose this implementation. This PR
also adds this capability to `synonym` token filters for use in search time
analyzers.

Relates to #29051

Commit:
Merge branch 'master' into reloadable-analyzers-feature

Commit:
Add documentation about reloading search analyzers (#42891)

elasticmachine commented:
Pinging @elastic/es-search

cbuescher commented:
@jpountz this PR merges the feature branch containing #42888 and accompanying documentation from #42891 into master. Those already got reviewed, but you mentioned you want to take another look at the whole branch.

jpountz commented on server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java:
```diff
@@ -0,0 +1,164 @@
+
+    @Override
+    public void close() {
+        storedComponents.close();
```
this should call super.close() too

jpountz commented on server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java:
```diff
@@ -0,0 +1,164 @@
+        @Override
+        public TokenStreamComponents getReusableComponents(Analyzer analyzer, String fieldName) {
+            ReloadableCustomAnalyzer custom = (ReloadableCustomAnalyzer) analyzer;
+            AnalyzerComponents components = custom.getStoredComponents();
```
nit: I got confused a bit while reading this logic because the variable is called components, but is actually the `Analyzer#getStoredComponents` rather than `Analyzer#getComponents`. Can we rename to `storedComponents`?

jpountz commented on server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java:
```diff
@@ -0,0 +1,164 @@
+            ReloadableCustomAnalyzer custom = (ReloadableCustomAnalyzer) analyzer;
+            AnalyzerComponents components = custom.getStoredComponents();
+            if (components == null || custom.shouldReload(components)) {
+                custom.setStoredComponents(custom.getComponents());
```
nit: we are accessing the components twice in the above calls (once via reload). In general I like extracting variables that may be updated by other threads to make things easier to reason about, could be extract the `Analyzer#getComponents` to a variable? This would guarantee that we are seeing exactly the same instance both times. This might not fit well with the current `shouldReload` method but maybe we can remove it and just do `components != storedComponents`?

jpountz reviewed (approved):
I left some minor comments but otherwise LGTM.

Commit:
Move mentions of updateable flag (#43536)

The new "updateable" flag used with synonym and synonym_graph should be
mentioned with the reload API endpoint which I also moved under the x-pack basic
role with this change.

Commit:
Merge branch 'master' into reloadable-analyzers-feature

Commit:
Moving reload_analyzers endpoint to xpack (#43559)

cbuescher commented on docs/reference/indices/reload-analyzers.asciidoc:
```diff
@@ -0,0 +1,65 @@
+
+[source,js]
+--------------------------------------------------
+POST /my_index/_reload_search_analyzers
```
This should be the same now after recent docs changes.

cbuescher commented on rest-api-spec/src/main/resources/rest-api-spec/api/indices.reload_search_analyzers.json:
```diff
@@ -0,0 +1,33 @@
+  "indices.reload_search_analyzers": {
+    "documentation": "https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-reload-analyzers.html",
+    "stability": "experimental",
+    "methods": ["GET", "POST"],
```
I think  I uses "indices.refresh" and "indices.flush" as blueprints here, they both support GET and POST. But happy to change if you think only POST is better.

Commit:
Addressing Adriens comments

cbuescher commented:
@jpountz I addressed your comments on this PR, also merged in a couple of other changes to the docs and moved the endpoint to xpack (both reviewed in #43536, #43559). Maybe you want to take another last look, the only thing I left like it was is allowing the GET/POST methods on the enpoint, but I'm happy to change that if you think its better. Let me know if I should change anything else, otherwise I'll try to get a green CI build before merging.

Commit:
Use correct nodeId in result

Commit:
Merge branch 'master' into reloadable-analyzers-feature

cbuescher commented:
@elasticmachine run elasticsearch-ci/2

cbuescher commented:
Talked to @jpountz and its okay to merge. If there are any changes that need to be done after the fact we can still get them in in a separate PR.

jpountz reviewed (commented):
LGTM.
